app/models/miq_server/at_startup.rb

```diff
@@ -3,13 +3,35 @@ module MiqServer::AtStartup
   include Vmdb::Logging
 
   module ClassMethods
+    def startup!
+      log_managed_entities
+      write_systemd_unit_files
+      clean_all_workers
+      clean_dequeued_messages
+      purge_report_results
+    end
+
     def log_managed_entities
       region = MiqRegion.my_region
       prefix = "#{_log.prefix} Region: [#{region.region}], name: [#{region.name}]"
       log_under_management(prefix)
       log_not_under_management(prefix)
     end
 
+    def write_systemd_unit_files
+      return unless MiqEnvironment::Command.supports_systemd?
+
+      _log.info("Writing Systemd unit files...")
+      MiqWorkerType.worker_class_names.each do |class_name|
+        worker_klass = class_name.safe_constantize
+        worker_klass.ensure_systemd_files if worker_klass&.systemd_worker?
```

Comment:
Am I reading this correctly?  We now only run this at startup as opposed to every `sync_workers` iteration of the `monitor_workers`loop?  I'm curious if this was a bug before that's getting fixed by this refactor.

Reply:
Right

There is a setting (`workers.worker_base.defaults.systemd_enabled`) which can be turned on/off per worker.  When this was first added it was in `sync_workers` so that this setting could take effect without restarting the server.

The check for `file_path.exist?` down here [:point_down:](https://github.com/ManageIQ/manageiq/pull/20824#discussion_r526937715) was needed because this was run so frequently.

Now that we need to change the contents of the files we can no longer use `.exist?` because it wouldn't update existing unit files.

So the way I saw it the options are either write out the files every `sync_workers` loop (terrible), check if the files are the same (still requires a `read()` every sync_workers loop or some kind of timestamp check), or move it out of `sync_workers` so it isn't run as often and we can just `write()` once and move all.

Now that this is the default for appliances I don't think the benefit of turning the setting on and off without restarting is actually a benefit, so I thought moving this to startup killed two birds with one stone.

Reply:
At this point I think even making systemd configurable per worker isn't a good idea and it should be a server-level setting.

Also long term I'd like the unit files to be installed by the RPM like they are for most software packages, though our workers are definitely more dynamic than most and we need to consider the provider pluggability aspect of that.

Reply:
Awesome, yeah, that makes sense to me.  It doesn't make sense to me to switch your worker process system on a live system without a restart.

Can you rewrite the commit that changed this logic to summarize this?  I'm afraid future you or I or someone else will see the refactoring commits and not realize why this logic changed.

Reply:
:+1: Absolutely, https://github.com/ManageIQ/manageiq/pull/20824/commits/f7d11bb9740b7a2c694a3ae160fee84a8dca6a9e

Reply:
😍  Future me or even Monday me appreciates it.  Thanks.